### PR TITLE
Focus frames with the mouse

### PR DESCRIPTION
**Pull request by t-wissmann**
This adds basic support for selecting frames with the mouse: either by
click or by mouse hover (if focus_follows_mouse is enabled). This change
only concerns the actual frame. If there is no frame shown (because
always_show_frame=off) or if there is a hole in the frame (because
frame_bg_transparent=on), then we only obtain events for the root window
that are currently ignored. (Interpreting event coordinates on the root
window is in the scope of another change)

On the implementation level, this introduces a static map that
transforms window ids into frame decorations (I think a 'framemanager'
is overkill here). The focusFrame() function is added to the root
window, because it affects the global state of hlwm (and I think
focus_client() should go to Root at some point, too).

This solves issue #20 (in principle) and adds a test case for it.

### CODE REVIEW

File: src/xmainloop.cpp [hunk 1]
```diff
@@ -334,6 +342,18 @@ void XMainLoop::enternotify(XCrossingEvent* ce) {
...
+        if (!c) {
+            // if it's not a client window, it's maybe a frame
+            FrameDecoration* frameDec = FrameDecoration::withWindow(ce->window);
+            if (frameDec) {
+                auto frame = frameDec->frame();
+                HSWeakAssert(frame);
+                if (frame) {
+                    root_->focusFrame(frame);
+                }
+            } else {
```
Comment (blarz) [inline src/xmainloop.cpp:354]: Is the empty else condition really necessary?

File: tests/test_basic_events.py [hunk 2]
```diff
@@ -50,3 +50,32 @@ def test_focus_follows_mouse(hlwm, mouse, focus_follows_mouse, single_floating):
...
+
+    # go to left frame
+    mouse.move_to(width * 0.25, height * 0.5)
+
+    # go to right frame
+    mouse.move_to(width * 0.75, height * 0.5)
+    if click:
+        mouse.click('1')
+
+    # we expect the focus to got to the second frame
```
Comment (blarz) [inline tests/test_basic_events.py:77]: s/got/go/

File: src/xmainloop.cpp (hunk 1 above)
Comment (t-wissmann) [reply src/xmainloop.cpp]: I've removed it (I had a printf debugging statement there before).

**Comment by Piping** [comment]
Hi,

Just found this awesome windows manager. Feels quite comfortable after an hour and trail and experiments.

Only thing I left un-configured is root windows access in X11. ( the mouse click behavior binding for root window)

Currently if I enabled this new build, I can only select the frame by clicking the frame border. ( if I enable hovering, and move mouse really fast, then the focus sometimes not working as well).

I assume it has to do with root windows access in the code right?  Is it possible to enable mouse binding to root windows now? or could transparent frame can catch mouse click like other windows that have the content?

**Comment by t-wissmann** [comment]
> Currently if I enabled this new build, I can only select the frame by clicking the frame border. ( if I enable hovering, and move mouse really fast, then the focus sometimes does not work).

Yes. That's probably because the frame only consists of the border of the frame.
As of now, hlwm does not handle clicks on the root window.

> I assume it has to do with root windows access in the code right? Is it possible to enable mouse binding to root windows now? or could transparent frame can catch mouse click like other windows that have the content?

I don't get your last sentence, but the reason for the lacking features is that nobody got time yet to implement it. The "frame_bg_transparent" setting is implemented by cutting an actual hole in the frame, so the frame window does not receive any click events if someone clicks into the hole, but the window behind receives the event instead.

**Comment by Piping** [comment]
> the reason for the lacking features is that nobody got time yet to implement it

@t-wissmann Cool. I would like to contribute code for the root windows access handles if it is not a philosophical design choice.

**Comment by t-wissmann** [comment]
> > the reason for the lacking features is that nobody got time yet to implement it
> 
> @t-wissmann Cool. I would like to contribute code for the root windows access handles if it is not a philosophical design choice.

OK great!

**Pull request merged**